map/search_api.cpp

```diff
@@ -222,6 +223,44 @@ bool SearchAPI::SearchInViewport(ViewportSearchParams const & params)
   return Search(p, false /* forceSearch */);
 }
 
+bool SearchAPI::SearchForDiscovery(DiscoverySearchParams const & params)
+{
+  CHECK(params.m_onResults, ());
+  CHECK(!params.m_query.empty(), ());
+  CHECK_GREATER(params.m_itemsCount, 0, ());
+
+  SearchParams p;
+  p.m_query = params.m_query;
+  p.m_inputLocale = "en";
+  p.m_viewport = params.m_viewport;
+  p.m_position = params.m_position;
+  p.m_maxNumResults = params.m_itemsCount;
+  p.m_mode = search::Mode::Viewport;
+  p.m_onResults = [params](Results const & results) {
+    if (!results.IsEndMarker())
+      return;
+
+    switch (params.m_sortingType)
+    {
+    case DiscoverySearchParams::SortingType::None:
+      params.m_onResults(results);
+      break;
+    case DiscoverySearchParams::SortingType::HotelRating:
+    {
+      Results r(results);
+      sort(r.begin(), r.end(), DiscoverySearchParams::HotelRatingComparator());
```

Comment:
Much better (and safer) is to implement following method in Results:
```c++
class Results {
  template <typename Fn>
  void SortBy(Fn && comparer)
  {
    sort(begin(), end(), forward<Fn>(comparer));
    for (...)
       r[i].SetPositionInResults(i);
  }
};
```

Reply:
After that you'll be able to test that in UNIT TESTS, and @mpimenov won't forget to fix this code after refactoring of results.

Reply:
Ok, nice idea